nexusclient/client.go

```diff
@@ -0,0 +1,725 @@
+	ResponseBody []byte
+}
+
+// Error implements the error interface.
+func (e *UnexpectedResponseError) Error() string {
+	if nexusapi.IsContentTypeJSON(e.Response.Header) {
+		var failure nexusapi.Failure
+		if err := json.Unmarshal(e.ResponseBody, &failure); err == nil && failure.Message != "" {
+			return fmt.Sprintf("%s: %s", e.Message, failure.Message)
+		}
+
+	}
+	return e.Message
+}
+
+func (c *Client) newUnexpectedResponseError(message string, response *http.Response) error {
+	body, err := io.ReadAll(response.Body)
+	if err != nil {
+		c.logger.Error("failed to read response body", "error", err)
+	}
+	return &UnexpectedResponseError{
+		Message:      message,
+		Response:     response,
+		ResponseBody: body,
+	}
+}
+
+// NewClient creates a new [Client] from provided [Options].
+// None of the options are required. Provide BaseServiceURL if you intend to use this client to make Nexus service calls
+// or leave empty when using this client only to deliver completions.
+func NewClient(options Options) (*Client, error) {
+	if options.HTTPClient == nil {
+		options.HTTPClient = http.DefaultClient
+	}
+	var serviceBaseURL *url.URL
+	if options.ServiceBaseURL != "" {
+		var err error
+		serviceBaseURL, err = url.Parse(options.ServiceBaseURL)
+		if err != nil {
+			return nil, err
+		}
+		if serviceBaseURL.Scheme != "http" && serviceBaseURL.Scheme != "https" {
+			return nil, ErrInvalidURLScheme
+		}
+	}
+	if options.LogHandler == nil {
+		options.LogHandler = slog.NewTextHandler(os.Stderr, &slog.HandlerOptions{Level: slog.LevelInfo})
+	}
+	if options.GetResultMaxRequestTimeout == 0 {
+		options.GetResultMaxRequestTimeout = time.Minute
+	}
+	if options.Marshaler == nil {
+		options.Marshaler = nexusapi.DefaultMarshaler
+	}
+
+	return &Client{
+		Options:        options,
+		serviceBaseURL: serviceBaseURL,
+		logger:         *slog.New(options.LogHandler),
+	}, nil
+}
+
+// A Handle used to cancel operations and get their result and status.
+// Must be explicitly closed.
+type OperationHandle struct {
+	operation string
+	id        string
+	state     nexusapi.OperationState
+
+	client *Client
+
+	// mutually exclusive with failure
+	response *http.Response
+	failure  *nexusapi.Failure
+}
+
+// Operation is the name of the operation this handle represents.
+func (h *OperationHandle) Operation() string {
+	return h.operation
+}
+
+// ID is the handler generated ID for this handle's operation.
+// Empty for synchronous operations.
+func (h *OperationHandle) ID() string {
+	return h.id
+}
+
+// State is the last known operation state.
+// Empty for handles created with [nexusclient.Client.GetHandle] before issuing a request to get the result.
+//
+// ⚠️ [nexusclient.OperationHandle.GetInfo] does not update the handle's State.
+func (h *OperationHandle) State() nexusapi.OperationState {
+	return h.state
+}
+
+// Close the handle's associated http response, if present.
+func (h *OperationHandle) Close() error {
```

Comment:
It's rough to require that a handle be closed. It's unclear to callers. It's worth discussing not using handles for synchronous responses.

Reply:
You'll need to close in async cases too.